File: backend/app/schemas/session_log.py
from datetime import date
from typing import Optional, List
from pydantic import BaseModel, field_validator


class SessionLogBase(BaseModel):
    member_id: int
    exercise_id: int
    date: Optional[date] = None
    sets: int
    reps: int
    weight_lbs: float
    duration_seconds: Optional[float] = None
    notes: Optional[str] = None

    @field_validator("sets", "reps")
    @classmethod
    def positive_int(cls, v: int) -> int:
        if v <= 0:
            raise ValueError("Must be positive")
        return v

    @field_validator("weight_lbs")
    @classmethod
    def non_negative_float(cls, v: float) -> float:
        if v < 0:
            raise ValueError("Weight must be non-negative")
        return v


class SessionLogCreate(SessionLogBase):
    pass


class SessionLogUpdate(BaseModel):
    date: Optional[date] = None
    sets: Optional[int] = None
    reps: Optional[int] = None
    weight_lbs: Optional[float] = None
    duration_seconds: Optional[float] = None
    notes: Optional[str] = None


class SessionLogRead(BaseModel):
    id: int
    member_id: int
    exercise_id: int
    date: date
    sets: int
    reps: int
    weight_lbs: float
    duration_seconds: Optional[float]
    notes: Optional[str]
    e1rm: float
    volume: float
    member_name: str
    exercise_name: str

    model_config = {"from_attributes": True}

File: frontend/src/types/sessionLog.ts
export interface SessionLog {
  id: number;
  member_id: number;
  exercise_id: number;
  date: string;
  sets: number;
  reps: number;
  weight_lbs: number;
  duration_seconds: number | null;
  notes: string | null;
  e1rm: number;
  volume: number;
  member_name: string;
  exercise_name: string;
}

export interface SessionLogCreate {
  member_id: number;
  exercise_id: number;
  date?: string;
  sets: number;
  reps: number;
  weight_lbs: number;
  duration_seconds?: number;
  notes?: string;
}

export interface SessionLogUpdate {
  date?: string;
  sets?: number;
  reps?: number;
  weight_lbs?: number;
  duration_seconds?: number;
  notes?: string;
}

export interface LogRow {
  exercise_id: number;
  exercise_name: string;
  category: string;
  tracking_type: 'weight_reps' | 'weight_duration';
  reps: string;
  duration: string;
  weight_lbs: string;
  notes: string;
  isDirty: boolean;
  placeholder_reps: string;
  placeholder_duration: string;
  placeholder_weight: string;
}
